athina_logger/tracing/trace.py

```diff
@@ -0,0 +1,174 @@
+        total_tokens: Optional[int] = None,
+        response_time: Optional[int] = None,
+        context: Optional[Dict] = None,
+        expected_response: Optional[str] = None,
+        custom_attributes: Optional[Dict] = None,
+        cost: Optional[float] = None,
+    ) -> Generation:
+        span = Generation(
+            name=name,
+            start_time=(start_time or datetime.datetime.utcnow()),
+            end_time=end_time,
+            span_type=span_type,
+            status=status,
+            attributes=attributes,
+            input=input,
+            output=output,
+            duration=duration,
+            version=version,
+            prompt=prompt,
+            response=response,
+            prompt_slug=prompt_slug,
+            language_model_id=language_model_id,
+            environment=environment,
+            functions=functions,
+            function_call_response=function_call_response,
+            tools=tools,
+            tool_calls=tool_calls,
+            external_reference_id=external_reference_id,
+            customer_id=customer_id,
+            customer_user_id=customer_user_id,
+            session_id=session_id,
+            user_query=user_query,
+            prompt_tokens=prompt_tokens,
+            completion_tokens=completion_tokens,
+            total_tokens=total_tokens,
+            response_time=response_time,
+            context=context,
+            expected_response=expected_response,
+            custom_attributes=custom_attributes,
+            cost=cost,
+        )
+        self._spans.append(span)
+        return span
+    
+    def to_dict(self):
+        trace_dict = self._trace.model_dump()
+        trace_dict["spans"] = [span.to_dict() for span in self._spans]
+        return trace_dict
+
+    def update(
+        self,
+        end_time: Optional[datetime.datetime] = None,
+        duration: Optional[int] = None,
+        status: Optional[str] = None,
+        attributes: Optional[Dict[str, Any]] = None,
+    ):
+        if end_time:
+            self._trace.end_time = end_time.utcnow().isoformat()
+        if status:
+            self._trace.status = status
+        if attributes:
+            self._trace.attributes = attributes
+        if duration:
+            self._trace.duration = duration
+
+    def end(self, end_time: Optional[datetime.datetime] = datetime.datetime.utcnow()):
+        try:
+            for span in self._spans:
+                span.end(end_time)
+            if self._trace.end_time is None:
+                self._trace.end_time = end_time.utcnow().isoformat()
+            if self._trace.duration is None:
+                self._trace.duration = int((end_time - datetime.datetime.fromisoformat(self._trace.start_time)).total_seconds())
+            request_dict = remove_none_values(self.to_dict())
+            RequestHelper.make_post_request(endpoint=f'{API_BASE_URL}/api/v1/trace/sdk', payload=request_dict, headers={
```

Comment:
Currently, this will be a blocking call I guess. Shall we use different module than requests which provides making async HTTP calls?

Reply:
Made the save call async for now
Yes, we should introduce a async http client for all our api calls.